File: leetcode/[0101]symmetric-tree.ts
// https://leetcode.com/problems/symmetric-tree/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSymmetric(root: TreeNode | null): boolean {
    return solution2(root);
}



// recursive dfs: 0 ms
function solution1(root: TreeNode | null): boolean {
    if (!root) return true;

    return f(root.left, root.right);
}
function f(left: TreeNode | null, right: TreeNode | null): boolean {
    if (!left && !right) return true;
    else if (!left || !right) return false;
    else if (left.val !== right.val) return false;

    return f(left.left, right.right) && f(left.right, right.left);
}



// iterative bfs: 1 ms
function solution2(root: TreeNode | null): boolean {
    if (!root) return true;

    const qu = [[root.left, root.right]];
    while (qu.length > 0) {
        const [left, right] = qu.shift()!;

        if (!left && !right) { }
        else if (!left || !right) return false;
        else if (left.val !== right.val) return false;
        else {
            qu.push(
                [left.left, right.right], [left.right, right.left]
            );
        }
    }

    return true;
}

File: leetcode/[0101]symmetric-tree.go
// https://leetcode.com/problems/symmetric-tree/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
	return solution2(root)
}



// iterative bfs: 0 ms
func solution1(root *TreeNode) bool {
	if root == nil {
		return true
	}

	type LR struct {
		L *TreeNode
		R *TreeNode
	}

	qu := []LR{}
	qu = append(qu, LR{root.Left, root.Right})

	for len(qu) > 0 {
		n := qu[0]
		qu = qu[1:]

		if n.L == nil && n.R == nil {
			// nothing
		} else if n.L == nil || n.R == nil {
			return false
		} else if n.L.Val != n.R.Val {
			return false
		} else {
			qu = append(qu, LR{n.L.Left, n.R.Right}, LR{n.L.Right, n.R.Left})
		}
	}

	return true
}



// recursive: 0 ms
func solution2(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return rec(root.Left, root.Right)
}
func rec(L *TreeNode, R *TreeNode) bool {
	if L == nil && R == nil {
		return true
	} else if L == nil || R == nil {
		return false
	} else if L.Val != R.Val {
		return false
	} else {
		return true && rec(L.Left, R.Right) && rec(L.Right, R.Left)
	}
}
